### PR TITLE
Caffe without the patch, cpp-package fixed also with Caffe plugin

### PR DESCRIPTION
**Pull request by cjolivier01**

### CODE REVIEW

File: cpp-package/example/lenet.cpp [hunk 1]
```diff
@@ -35,19 +35,19 @@ class Lenet {
         Convolution("conv1", data, conv1_w, conv1_b, Shape(5, 5), 20);
     Symbol tanh1 = Activation("tanh1", conv1, ActivationActType::tanh);
     Symbol pool1 = Pooling("pool1", tanh1, Shape(2, 2), PoolingPoolType::max,
-      false, PoolingPoolingConvention::valid, Shape(2, 2));
+      false, false, PoolingPoolingConvention::valid, Shape(2, 2));
```
Comment (cjolivier01) [inline cpp-package/example/lenet.cpp]: Whatshould the default here be? Cudnn

File: cpp-package/example/resnet.cpp [hunk 2]
```diff
@@ -43,7 +46,7 @@ Symbol getConv(const std::string & name, Symbol data,
                                   kernel, num_filter, stride, Shape(1, 1),
                                   pad, 1, 512);
 
-  Symbol bn = BatchNorm(name + "_bn", conv, 2e-5, bn_momentum, false);
+  Symbol bn = BatchNorm(name + "_bn", conv, BN_GAMMA, BN_BETA, 2e-5, bn_momentum, false);
```
Comment (cjolivier01) [inline cpp-package/example/resnet.cpp]: What is the desired default gamma and beta?

File: src/executor/graph_executor.cc [hunk 3]
```diff
@@ -905,8 +905,6 @@ GraphExecutor::CachedSegOpr GraphExecutor::CreateCachedSegOpr(size_t topo_start,
     // you need to copy it out. (potential memory leak risk)
     char *p_opr_name = new char[opr_names.size() + 1];
     memcpy(p_opr_name, opr_names.c_str(), opr_names.size() + 1);
-#else
```
Comment (cjolivier01) [inline src/executor/graph_executor.cc]: This isn't necessary and generates a warning

**Comment by cjolivier01** [comment]
Fixing lint...

**Comment by cjolivier01** [comment]
travis appears to be broken (nohup)

File: cpp-package/include/mxnet-cpp/kvstore.hpp [hunk 4]
```diff
@@ -14,8 +14,8 @@
 #include "mxnet-cpp/kvstore.h"
 #include "mxnet-cpp/optimizer.h"
 
-#ifndef KVSTORE_HPP
-#define KVSTORE_HPP
+#ifndef CPP_PACKAGE_INCLUDE_MXNET_CPP_KVSTORE_HPP_
```
Comment (cjolivier01) [inline cpp-package/include/mxnet-cpp/kvstore.hpp]: These are lint fixes which captured problems in the original cpp-package headers

File: cpp-package/example/charRNN.cpp [hunk 5]
```diff
@@ -199,7 +199,7 @@ class BucketSentenceIter : public DataIter {
  public:
   BucketSentenceIter(string filename, int minibatch, Context context) : batch(minibatch),
   current(-1), device(context) {
-    auto& content = readContent(filename);
+    auto content = readContent(filename);
```
Comment (cjolivier01) [inline cpp-package/example/charRNN.cpp]: This was incorrectly taking a reference of a temporary

File: cpp-package/example/charRNN.cpp [hunk 6]
```diff
@@ -343,7 +343,8 @@ class BucketSentenceIter : public DataIter {
         chars.push_back(c);
       }
     }
-    return {map, chars};
+    // Note: Can't use {} because this would hit the explicit constructor
+    return tuple<unordered_map<wchar_t, mx_float>, vector<wchar_t>>(map, chars);
```
Comment (cjolivier01) [inline cpp-package/example/charRNN.cpp]: My gcc-5.4 won't allow this

File: Makefile [hunk 7]
```diff
@@ -133,7 +138,7 @@ endif
 .PHONY: clean all test lint doc clean_all rcpplint rcppexport roxygen\
 	cython2 cython3 cython cyclean
 
-all: lib/libmxnet.a lib/libmxnet.so $(BIN)
+all: lib/libmxnet.a lib/libmxnet.so $(BIN) cpp-package-example-all
```
Comment (piiswrong) [inline Makefile]: shouldn't build cpp-package by default

File: cpp-package/include/mxnet-cpp/kvstore.hpp (hunk 4 above)
Comment (piiswrong) [inline cpp-package/include/mxnet-cpp/kvstore.hpp]: you can make lint ignore CPP_PACKAGE_INCLUDE

**Comment by piiswrong** [comment]
@lx75249

File: cpp-package/example/resnet.cpp (hunk 2 above)
Comment (lx75249) [inline cpp-package/example/resnet.cpp]: Gamma and beta are learnable parameters that can be initialized automatically by initializer, but op.h genereator can't distinguish parameters from hyperparameters currently, as mxnet doesn't provide such information. In older versions, gamma and beta were not declared explicitly. We can use `Operator("BatchNorm")` here to avoid specifying gamma and beta before figuring out how to detect parameters.

**Comment by piiswrong** [comment]
I think we just need to put add arguments for beta and gamma

Xin Li <notifications@github.com>于2017年3月26日 周日上午12:15写道：

File: cpp-package/include/mxnet-cpp/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+# Rebuildable file(s)
+op.h
```
Comment (lx75249) [inline cpp-package/include/mxnet-cpp/.gitignore]: This will force everyone to build the whole library..

File: Makefile (hunk 7 above)
Comment (lx75249) [inline Makefile]: It only generates a header file for cpp-package, and is necessary since updates that break cpp-package will be reflected in the test..

File: cpp-package/include/mxnet-cpp/kvstore.hpp (hunk 4 above)
Comment (cjolivier01) [inline cpp-package/include/mxnet-cpp/kvstore.hpp]: Ok. I went ahead and fixed it so that we could have some clean, homogenous code.

File: Makefile (hunk 7 above)
Comment (cjolivier01) [inline Makefile]: are cpp-package tests run during CI builds? I think they they probably aren't because they didn't build at all. In the meantime, how will we keep people from "breaking" these compiliables again?

File: cpp-package/include/mxnet-cpp/.gitignore (hunk 8 above)
Comment (cjolivier01) [inline cpp-package/include/mxnet-cpp/.gitignore]: You mean to build libmxnet.so?  I would expect so, yes.  The problem I encountered is that this doesn;t all normally get built, so when I went ot build it, it didn't build at all, and I spent 3 days fixing it.  At least by building all or part of it as part of the build, we can ensure that the main codebase doesn't diverge again.  It's a quick build (cpp-package)
wdyt?

Comment (cjolivier01) [inline cpp-package/include/mxnet-cpp/.gitignore]: scala and R also require libmxnet.so to be built, btw.

File: Makefile (hunk 7 above)
Comment (cjolivier01) [inline Makefile]: How about we just build lenet by default?  At any rate, the build for all of them is quick -- a few seconds.

Comment (cjolivier01) [inline Makefile]: Or you mean a config.mk flag and w can turn it on for the CI build? That might be ok.

File: cpp-package/include/mxnet-cpp/.gitignore (hunk 8 above)
Comment (cjolivier01) [inline cpp-package/include/mxnet-cpp/.gitignore]: I am a little confused. The expectation is that someone will check out the mxnet source tree and include directly from mxnet/cpp-package/include, but dynamically load the libmxnet.so that came with their pip install rather than one that they built locally?  That seems unlikely and dangerous, correct?  
The cmake build "distributes" mxnet headers during a build if INSTALL_INCLUDE_DIR is defined as a path, although not yet for cpp-package.  If we're going to make a distribution like apt-get install mxnet-dev, I would expect that we can deploy the built op.h

wdyt?

File: Makefile (hunk 7 above)
Comment (lx75249) [inline Makefile]: If someone changes operator api, cpp-package test will fail since op.h changes after building cpp-package.

File: cpp-package/include/mxnet-cpp/.gitignore (hunk 8 above)
Comment (lx75249) [inline cpp-package/include/mxnet-cpp/.gitignore]: yes you're right...

**Comment by piiswrong** [comment]
@lx75249 @cjolivier01 Is this good for merging?

**Comment by cjolivier01** [comment]
Fixing build things. Also, need to do Operator("BatchNorm")

**Comment by cjolivier01** [comment]
other checkin of ft.cc contrib operator broke the build on Windows, libcufft library not linked. Added to cmake.
